File: leetcode/atoi/atoi.java
package atoi;

class Solution {
    public int myAtoi(String s) {
        if(s.isEmpty()) return 0;

        int i = 0;

        while(i < s.length() && s.charAt(i) == ' ') i++;

        boolean isNegative = false;

        if(i < s.length() && (s.charAt(i) == '-' || s.charAt(i) == '+')) {
            isNegative = s.charAt(i) == '-';
            i++;
        }

        while(i < s.length() && s.charAt(i) == '0') i++;

        String ansStr = "";
        int count = 0;
        while(i < s.length() && count < 11 && '0' <= s.charAt(i) && s.charAt(i) <= '9') {
            ansStr += s.charAt(i++);
            count++;
        }

        if(ansStr.isEmpty()) return 0;

        try {
            int ans = isNegative ? Integer.valueOf("-" + ansStr) : Integer.valueOf(ansStr);
            return ans;
        } catch (Exception e) {
            return isNegative ? Integer.MIN_VALUE : Integer.MAX_VALUE;
        }
    }
}

File: leetcode/atoi/atoi.cpp
/*
 * Time Complexity: O(N)
 *   - N is the length of the input string.
 *   - Each character is processed at most once.
 *
 * Space Complexity: O(1)
 *   - Only a constant amount of extra space is used (excluding input and output).
 */

#include <string>
#include <stdexcept>
#include <limits.h>

using std::string;
using std::exception;

class Solution {
public:
    int myAtoi(string s) {
        if(s.length() == 0) return 0;

        int i = 0;
        while(i < s.length() && s[i] == ' ') i++;

        bool isNegative = false;
        string ansStr = "";
        if(i < s.length() && (s[i] == '+' || s[i] == '-')) {
            isNegative = s[i] == '-';
            if(isNegative) ansStr += "-";
            i++;
        }

        while(i < s.length() && s[i] == '0') i++;

        int count = 0;
        while(i < s.length() && count < 11 && '0' <= s[i] && s[i] <= '9') {
            ansStr += s[i++];
            count++;
        }

        if(ansStr.length() == 0 || (isNegative && ansStr.length() == 1)) return 0;

        try {
            int ans = stoi(ansStr);
            return ans;
        } catch (exception& e) {
            return isNegative ? INT_MIN : INT_MAX;
        }
    }
};